File: src/03.py
import numpy as np
import timeit

def count_trees(maze, dx, dy):
    x_lim = len(maze)
    y_lim = len(maze[0])
    x = dx
    y = dy
    num_trees = 0
    while x < x_lim:
        if maze[x][y % y_lim] == '#':
            num_trees += 1
        x += dx
        y += dy
    return num_trees


def main():
    with open("../input/03.txt") as f:
        maze = f.readlines()
    maze = [s.strip() for s in maze]
    check_list = [
        [1, 1],
        [1, 3],
        [1, 5],
        [1, 7],
        [2, 1],
    ]
    timeit_num = 1000
    print(
        timeit.timeit(lambda: count_trees(maze, 1, 3), number=timeit_num)
        / timeit_num)
    print(count_trees(maze, 1, 3))
    print(timeit.timeit(
        lambda: np.product([count_trees(maze, *l) for l in check_list]),
        number=1000,
    ) / timeit_num)
    print(np.product([count_trees(maze, *l) for l in check_list]))

main()

File: src/03.jl
#!/usr/bin/julia

function count_trees(maze::AbstractArray{Bool, 2}, dx::Integer, dy::Integer)
    x_lim, y_lim = size(maze)
    x = dx
    y = dy
    num_trees = 0
    while x < x_lim
        if maze[begin + x, begin + y]
            num_trees += 1
        end
        x += dx
        y += dy
        y %= y_lim
    end
    return num_trees
end


function main()
    lines = open("../input/03.txt", "r") do io
        map(strip, readlines(io))
    end
    maze = hcat(Array(l) for l in lines) .== '#'
    maze = transpose(maze)

    check_list = [
        [1, 1],
        [1, 3],
        [1, 5],
        [1, 7],
        [2, 1],
    ]

    @time count_trees(maze, 1, 3)
    @timev count_trees(maze, 1, 3)
    @info "Found answer" count_trees(maze, 1, 3)
    @timev prod(count_trees(maze, l...) for l in check_list)
    @info "Found answer" prod(count_trees(maze, l...) for l in check_list)
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
